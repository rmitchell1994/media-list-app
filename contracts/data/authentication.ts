export type UserPassword = {
  password: string;
};

export interface UserInfo extends UserPassword {
  email: string;
}

export type UpdateAccountState = {
  error: boolean;
  success: boolean;
  isLoading: boolean;
  isAccountDeleted: boolean;
};

export type UpdateAccountAction = {
  type:
    | "ACCOUNT_UPDATE_SUCCESS"
    | "ACCOUNT_UPDATE_ERROR"
    | "ACCOUNT_UPDATE_REQUEST"
    | "ACCOUNT_UPDATE_DELETE";
};
