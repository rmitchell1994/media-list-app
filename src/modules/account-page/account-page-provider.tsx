import React, {
  createContext,
  Dispatch,
  FC,
  ReactNode,
  useReducer,
} from "react";
import {
  UpdateAccountState,
  UpdateAccountAction,
} from "../../../contracts/data/authentication";
import { accountPageReducer, initialState } from "./account-page-reducer";

type Props = {
  children: ReactNode;
};

export const AccountPageContext = createContext<{
  status: UpdateAccountState;
  dispatch: Dispatch<UpdateAccountAction>;
}>({ status: initialState, dispatch: () => null });

const AccountPageProvider: FC<Props> = ({ children }: Props) => {
  const [status, dispatch] = useReducer(accountPageReducer, initialState);

  return (
    <AccountPageContext.Provider value={{ dispatch, status }}>
      {children}
    </AccountPageContext.Provider>
  );
};

export default AccountPageProvider;
