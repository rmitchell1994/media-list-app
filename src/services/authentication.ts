import firebase from "firebase/app";
import "firebase/auth";
import { UserInfo } from "../../contracts/data/authentication";

export const createEmailUser = async ({
  email,
  password,
}: UserInfo): Promise<firebase.auth.UserCredential> =>
  firebase.auth().createUserWithEmailAndPassword(email, password);

export const signInUser = async ({
  email,
  password,
}: UserInfo): Promise<firebase.auth.UserCredential> =>
  firebase.auth().signInWithEmailAndPassword(email, password);

export const signOut = async (): Promise<void> => firebase.auth().signOut();
