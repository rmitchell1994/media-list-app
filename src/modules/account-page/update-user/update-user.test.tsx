import React from "react";
import { useMutation } from "react-query";
import { act, fireEvent, render, screen } from "../../../../test/test-utils";
import UpdateUser from "./update-user";
import { AccountPageContext } from "../account-page-provider";
import { updateUserPassword } from "../../../services/authentication";

jest.mock("../../../services/authentication", () => ({
  updateUserPassword: jest.fn().mockResolvedValue(() => Promise.resolve()),
}));

jest.mock("react-query", () => ({
  ...jest.requireActual<{ moduleName: "react-query" }>("react-query"),
  useMutation: jest.fn().mockReturnValue({
    mutate: jest.fn(),
  }),
}));

describe("Update User Component", () => {
  const dispatch = jest.fn();
  const status = {
    isLoading: false,
    error: false,
    success: false,
  };

  beforeEach(() => {
    jest.restoreAllMocks();
  });

  it("should render form", () => {
    const { asFragment } = render(
      <AccountPageContext.Provider value={{ dispatch, status }}>
        <UpdateUser />
      </AccountPageContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call useMutation and update user password", () => {
    render(
      <AccountPageContext.Provider value={{ dispatch, status }}>
        <UpdateUser />
      </AccountPageContext.Provider>
    );

    act(() => {
      fireEvent.change(screen.getByLabelText("Update Password"), {
        target: { value: "password" },
      });
      fireEvent.click(screen.getByText("Submit"));
    });

    expect(useMutation).toHaveBeenCalled();
    expect(useMutation).toHaveBeenCalledWith(updateUserPassword, {
      onMutate: expect.any(Function),
      onError: expect.any(Function),
      onSuccess: expect.any(Function),
    });
  });
});
