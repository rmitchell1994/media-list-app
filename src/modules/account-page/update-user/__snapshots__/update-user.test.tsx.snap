// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update User Component should render form 1`] = `
<DocumentFragment>
  <div
    data-testid="update-user"
  >
    <form>
      <label>
        Update Password
        <input
          data-testid="update-password-input"
          name="password"
          placeholder="New Password"
          type="password"
        />
      </label>
      <button
        data-testid="update-password-submit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
